core/src/main/java/com/linecorp/armeria/common/stream/PathStreamMessage.java

```diff
@@ -0,0 +1,324 @@
+import com.linecorp.armeria.common.util.Exceptions;
+import com.linecorp.armeria.internal.common.stream.NoopSubscription;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.util.ReferenceCountUtil;
+import io.netty.util.concurrent.EventExecutor;
+
+final class PathStreamMessage implements StreamMessage<HttpData> {
+
+    static final int DEFAULT_FILE_BUFFER_SIZE = 4096;
+
+    private final CompletableFuture<Void> completionFuture = new EventLoopCheckingFuture<>();
+
+    private final Path path;
+    private final ByteBufAllocator alloc;
+    private final int bufferSize;
+
+    private boolean subscribed;
+
+    @Nullable
+    private volatile PathSubscription pathSubscription;
+
+    PathStreamMessage(Path path, ByteBufAllocator alloc, int bufferSize) {
+        this.path = requireNonNull(path, "path");
+        this.alloc = requireNonNull(alloc, "alloc");
+        this.bufferSize = bufferSize;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return !completionFuture.isDone();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        if (isOpen()) {
+            return false;
+        }
+        final PathSubscription pathSubscription = this.pathSubscription;
+        return pathSubscription == null || pathSubscription.position == 0;
+    }
+
+    @Override
+    public long demand() {
+        final PathSubscription pathSubscription = this.pathSubscription;
+        if (pathSubscription != null) {
+            return pathSubscription.requested;
+        } else {
+            return 0;
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return completionFuture;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super HttpData> subscriber, EventExecutor executor) {
+        subscribe(subscriber, executor, EMPTY_OPTIONS);
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super HttpData> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+        if (executor.inEventLoop()) {
+            subscribe0(subscriber, executor, options);
+        } else {
+            executor.execute(() -> subscribe0(subscriber, executor, options));
+        }
+    }
+
+    private void subscribe0(Subscriber<? super HttpData> subscriber, EventExecutor executor,
+                            SubscriptionOption... options) {
+        if (subscribed) {
+            subscriber.onSubscribe(NoopSubscription.get());
+            subscriber.onError(new IllegalStateException("Only single subscriber is allowed!"));
+            return;
+        }
+        subscribed = true;
+
+        final AsynchronousFileChannel fileChannel;
+        try {
+            fileChannel = AsynchronousFileChannel.open(path, StandardOpenOption.READ);
```

Comment:
Question: I don't remember how `AsynchronousFileChannel` works exactly under the hood.
Can we specify the `executor` when opening?

Reply:
There is [an open method](https://docs.oracle.com/javase/8/docs/api/java/nio/channels/AsynchronousFileChannel.html#open-java.nio.file.Path-java.util.Set-java.util.concurrent.ExecutorService-java.nio.file.attribute.FileAttribute...-
) for specifying an `ExecutorService`. 

`AsynchronousFileChannel` internally blocks IO thread for reading data from disk on Linux.
https://stackoverflow.com/questions/39501924/java-asynchronousfilechannel-thread-usage
It seems better to set Armeria's blocking task executor.